crates/primitives/starknet/src/fees/mod.rs

```diff
@@ -0,0 +1,251 @@
+    let (n_modified_contracts, n_modified_keys, n_class_updates) = state.count_state_changes();
+    // TODO: consider the `validate` resources.
+    // FIXME: https://github.com/keep-starknet-strange/madara/issues/227
+    // The `&None` corresponds to the `validate` execution for now.
+    let non_optional_call_infos: Vec<&CallInfo> = vec![execute_call_info, &None].into_iter().flatten().collect();
+    let mut l2_to_l1_payloads_length = vec![];
+    for call_info in non_optional_call_infos {
+        l2_to_l1_payloads_length.extend(
+            call_info
+                .get_sorted_l2_to_l1_payloads_length()
+                .map_err(|_| TransactionExecutionErrorWrapper::FeeComputationError)?,
+        );
+    }
+    let l1_gas_usage = calculate_tx_gas_usage(
+        &l2_to_l1_payloads_length,
+        n_modified_contracts,
+        n_modified_keys + usize::from(FEE_TRANSFER_N_STORAGE_CHANGES_TO_CHARGE),
+        None,
+        n_class_updates,
+    );
+    // Add additional Cairo resources needed for the OS to run the transaction.
+    let total_vm_usage = &execution_resources.vm_resources
+        + &get_additional_os_resources(execution_resources.syscall_counter.clone(), tx_type.into())
+            .map_err(|_| TransactionExecutionErrorWrapper::FeeComputationError)?;
+    let total_vm_usage = total_vm_usage.filter_unused_builtins();
+    let mut tx_resources = BTreeMap::from([
+        (GAS_USAGE.to_string(), l1_gas_usage),
+        (N_STEPS_RESOURCE.to_string(), total_vm_usage.n_steps + total_vm_usage.n_memory_holes),
+    ]);
+    tx_resources.extend(total_vm_usage.builtin_instance_counter);
+    Ok(tx_resources)
+}
+
+/// Charges the fees for a specific execution resources.
+///
+/// # Arguments
+///
+/// * `state` - State object to get the state changes.
+/// * `block_context` - Block context to get information needed to compute the fees.
+/// * `account_tx_context` - Account context.
+/// * `resources` - Execution resources.
+///
+/// # Returns
+///
+/// * [Fee] - Amount charged for the transaction.
+/// * [Option<CallInfo>] - Call info of the fee transfer tx.
+///
+/// # Errors
+///
+/// [TransactionExecutionErrorWrapper] if any step of the fee transfer computation/transaction
+/// fails.
+pub fn charge_fee<S: State + StateChanges>(
+    state: &mut S,
+    block_context: &BlockContext,
+    account_tx_context: &AccountTransactionContext,
+    resources: &BTreeMap<String, usize>,
+) -> Result<(Fee, Option<CallInfo>), TransactionExecutionErrorWrapper> {
+    let no_fee = Fee::default();
+    if account_tx_context.max_fee == no_fee {
+        // Fee charging is not enforced in some tests.
+        return Ok((no_fee, None));
+    }
+
+    let actual_fee = calculate_tx_fee(resources, block_context)
+        .map_err(|_| TransactionExecutionErrorWrapper::FeeComputationError)?;
+    let fee_transfer_call_info =
+        execute_fee_transfer(state, &mut ExecutionResources::default(), block_context, account_tx_context, actual_fee)?;
```

Comment:
`ExecutionResources::default` why do we use the default value ?

Reply:
because we want a new empty execution resource to run the fee transfer